keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,132 @@
+
+### Goals
+- Provide more strong guarantees around memory availability for guaranteed pods
+- Provide (best effort) guarantees around memory availability for burstable pods
+- Make use of new cgroup v2 memory controller settings such as `memory.min`, `memory.low` for pod and container cgroups
+
+### Non-Goals
+- additional qos design
+- support other resources qos
+
+## Proposal
+Use memory controller interfaces of cgroups v2 to support memory qos of pod qos class.
+
+Kubernetes cgroup manager uses `memory.limit_in_bytes` in cgroups v1 and `memory.max` in cgroups v2 to limit memory capacity and uses `oom_scores` to determine order of killing container process when OOM occurs.
+
+Above implement has following disadvantages:
+- `Guaranteed` pod memory can not be fully reserved, page cache is at risk of being recycled
+- `Burstable` pod memory allocation is not well protected, there may occur allocation latency frequently when node memory nearly runs out
+
+Currently we only use `memory.limit_in_bytes = sum(pod.spec.containers.resources.limits[memory])` to limit memory hardly in cgroups v1 and `memory.max = sum(pod.spec.containers.resources.limits[memory])` in cgroups v2. And `resources.requests[memory]` is not yet used neither cgroups v1 nor cgroups v2.
+
+cgroups v2 introduces a better way to limit and guarantee memory quality.
+
+| File | Description |
+| -------- | -------- |
+| memory.high | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. |
+| memory.max | memory.max is the memory usage hard limit, acting as the final protection mechanism: If a cgroup's memory usage reaches this limit and can't be reduced, the system OOM killer is invoked on the cgroup. Under certain circumstances, usage may go over the memory.high limit temporarily. When the high limit is used and monitored properly, memory.max serves mainly to provide the final safety net. The default is max. |
+| memory.low | memory.low is the best-effort memory protection, a "soft guarantee" that if the cgroup and all its descendants are below this threshold, the cgroup's memory won't be reclaimed unless memory can’t be reclaimed from any unprotected cgroups. |
+| memory.min | memory.min specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup's memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked. |
+
+This proposal uses `memory.min` to retain request memory for `Guaranteed` pods and containers. And for `Burstable` pods, it uses `memory.low` to achieve best-effort memory protection at pod level.
```

Comment:
is there any plan for exposing `memory.high`?

Reply:
I think that using a different limit depending on the container class could be a bit confusing.  It will make more difficult in future to expose all the cgroups knobs.  It is valid for a cgroup to set all of the `memory.[min,low,high,max]` values.

I'd always map the memory request to `memory.min` and add two new knobs to map `memory.low` and `memory.high`.  In this way we have the possibility to set the 4 different levels.

What do you think?

Reply:
Do you mean we should set memory.min = request both for guaranteed and burstable?

And in this implementation, what about memory low and high? Add new qos class for that?

Reply:
@giuseppe I think different logic should be wrapped for different qos class. For example, guaranteed pod memory must be strongly protected to ensure that memory is always allocated with low latency. For burstable pod, try to keep memory as much as possible, but in extreme cases, it can also be reclaimed. At this level, corresponding to the oom_score logic, burstable will also be killed. That's why I set guaranteed request to memory.min, and burstable request to memory.low.

Reply:
On the other hand, exposing too many cgroups memory controller interfaces to users will increase the cost for understanding, and it is difficult to harmonize with the existing concepts of qos class. Of course, I think it's a good choice to expose memory.high in a new way.

Reply:
one of the main motivations of moving to cgroup v2 is take advantage of these additional memory knobs.

If we pick `memory.min` or `memory.low` depending on the QoS class, then it will be more difficult to expose the other setting (how should it be called?).

I am fine if we don't cover the new knobs as part of this KEP, but we should at least imagine a way to support them in future and choosing `memory.min` or `memory.low` depending on the QoS, IMO, makes it more difficult

Reply:
> one of the main motivations of moving to cgroup v2 is take advantage of these additional memory knobs.
> 

If this is the motive, I basically agree with you. Binding to qos class does make further extensions a little more difficult.

> If we pick `memory.min` or `memory.low` depending on the QoS class, then it will be more difficult to expose the other setting (how should it be called?).
> 

I am ok with setting `memory.min = memory request` in this proposal no matter what qos class. And we can next discuss new knobs(memory.min / memory.low), maybe in another proposals. Is it good for you?

Reply:
@bobbypage @mrunalp What do you think?

Reply:
> > 
> 
> I am ok with setting `memory.min = memory request` in this proposal no matter what qos class. And we can next discuss new knobs(memory.min / memory.low), maybe in another proposals. Is it good for you?

thanks.  Yes that definitely works for me!

Just, is it better to map it to `memory.min` or `memory.low`? :smile:

Reply:
In order to make the strategy less aggressive and ensure both system and Kubernetes reserved memory, can we refer to the following rules:
- `memory.low` for memory requests no mater what qos class is
- `memory.min` for kube-reserved & system-reserved